linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala

```diff
@@ -1,46 +1,232 @@
+        val deadline = Time.fromNanoseconds(values(1).toLong)
+        FDeadline(timestamp, deadline)
+      }
+
+      def get(headers: HeaderMap): Option[FDeadline] =
+        headers.getAll(Key).foldLeft[Option[FDeadline]](None) { (d0, v) =>
+          (d0, read(v).toOption) match {
+            case (Some(d0), Some(d1)) => Some(FDeadline.combined(d0, d1))
+            case (d0, d1) => d0.orElse(d1)
+          }
+        }
+
+      def write(d: FDeadline): String =
+        s"${d.timestamp.inNanoseconds} ${d.deadline.inNanoseconds}""
+
+      def set(headers: HeaderMap, deadline: FDeadline): Unit = {
+        val _ = headers.set(Key, write(deadline))
+      }
+
+      def clear(headers: HeaderMap): Unit = {
+        val _ = headers.remove(Key)
+      }
+
+      /**
+       * Extract the deadline from the request and, if it exists, use
+       * either that deadline or the one already on the local context,
+       * whichever is stricter.
+       */
+      class ServerFilter extends SimpleFilter[Request, Response] {
+        def apply(req: Request, service: Service[Request, Response]) =
+          get(req.headerMap) match {
+            case None => service(req)
+            case Some(reqDeadline) =>
+              clear(req.headerMap)
+              val deadline = FDeadline.current match {
+                case None => reqDeadline
+                case Some(current) => FDeadline.combined(reqDeadline, current)
+              }
+              Contexts.broadcast.let(FDeadline, deadline) {
+                service(req)
+              }
+          }
+      }
+
+      class ClientFilter extends SimpleFilter[Request, Response] {
+        def apply(req: Request, service: Service[Request, Response]) =
+          FDeadline.current match {
+            case None => service(req)
+            case Some(deadline) =>
+              set(req.headerMap, deadline)
+              service(req)
+          }
+      }
+    }
+
+    object Dtab {
+      val UserKey = Headers.Prefix + "dtab"
+      val CtxKey = Ctx.Prefix + "dtab"
+      private val EmptyReturn = Return(FDtab.empty)
+
+      def get(headers: HeaderMap, key: String): Try[FDtab] =
+        if (!headers.contains(key)) EmptyReturn
+        else Try { FDtab(headers.getAll(key).view.flatMap(FDtab.read(_)).toIndexedSeq) }
+
+      def get(headers: HeaderMap): Try[FDtab] =
+        for {
+          ctx <- get(headers, CtxKey)
+          user <- get(headers, UserKey)
+        } yield ctx ++ user
+
+      def clear(headers: HeaderMap): Unit = {
+        val _c = headers.remove(CtxKey)
+        val _u = headers.remove(UserKey)
+      }
+
+      def set(dtab: FDtab, msg: Message): Unit =
+        if (dtab.nonEmpty) {
+          val _ = msg.headerMap.set(CtxKey, dtab.show)
+        }
+
+      /**
+       * Extract a Dtab from the L5d-Ctx-Dtab and L5d-Dtab headers (in
+       * that order) and append them to the local context.
+       *
+       * The L5d-Ctx-Dtab header is intended to be set by a linkerd
```

Comment:
Yeah i understand.  I think this is simple because we can clarify the context of this header more clearly:

l5d-ctx headers are ONLY to be generated by linkerd. They should always be stripped at the edge.  l5d-dtab and l5d-sample are ONLY to be set by users, and only admitted from trusted sources.

In the same way that l5d-sample influences l5d-ctx-trace, l5d-dtab influences l5d-ctx-dtab.  l5d-ctx headers should be totally opaque to users and services.